[visionOS] Gaze glow shape is incorrect for elements with associated label and non-uniform border radii or clip path

lilyspiniolas opened the pull request:
#### b1e40d13effbbf84d58e2b9a575aff21ade320cb
<pre>
[visionOS] Gaze glow shape is incorrect for elements with associated label and non-uniform border radii or clip path
<a href="https://bugs.webkit.org/show_bug.cgi?id=294952">https://bugs.webkit.org/show_bug.cgi?id=294952</a>
<a href="https://rdar.apple.com/154258426">rdar://154258426</a>

Reviewed by Abrar Rahman Protyasha.

During the shrink wrapping of interaction regions, we previously
collected the rect of each discovered region, and then set the
clip path equal to the result of `pathWithShrinkWrappedRects`
with the collected rects passed as a parameter.

Now, for each discovered region, we only add its rect to the
collection if the region does not have a clip path. If a clip
path is present, we add it a collection of clip paths. The
final clip path is then set to a path consisting of all of the
discovered clip paths + the result of `pathWithShrinkWrappedRect`
for the collected rects.

* LayoutTests/interaction-region/clip-path-with-label-expected.txt: Added.
* LayoutTests/interaction-region/clip-path-with-label.html: Added.
* Source/WebCore/rendering/EventRegion.cpp:
(WebCore::EventRegionContext::shrinkWrapInteractionRegions):

Canonical link: <a href="https://commits.webkit.org/296646@main">https://commits.webkit.org/296646@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5685983e0a554c8e5f94667579bed7d40b497f75

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/109147 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/28805 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/19232 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/114355 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/59434 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/111110 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/29486 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/37429 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/82944 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/112095 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/23460 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/98308 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/63391 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/22857 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/16450 "Passed tests") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/59016 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/92831 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/16493 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/117470 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/36191 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/26798 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/91962 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/36563 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/94571 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/91768 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23371 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/36685 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/14437 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/32052 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/36087 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/41589 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/35778 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/39115 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/37465 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WebCore/rendering/EventRegion.cpp:
```diff
@@ -313,18 +317,38 @@ void EventRegionContext::shrinkWrapInteractionRegions()
+            Path path;
+            path.addPath(shrinkWrappedRects, transform);
+            for (Path clipPath : discoveredClipPaths)
+                path.addPath(clipPath, transform);
```
Nit: you could get away with r-value `AffineTransform`s, i.e. `addPath(rect, { })`

lilyspiniolas committed:
[visionOS] Gaze glow shape is incorrect for elements with associated label and non-uniform border radii or clip path
https://bugs.webkit.org/show_bug.cgi?id=294952
rdar://154258426

Reviewed by Abrar Rahman Protyasha.

During the shrink wrapping of interaction regions, we previously
collected the rect of each discovered region, and then set the
clip path equal to the result of `pathWithShrinkWrappedRects`
with the collected rects passed as a parameter.

Now, for each discovered region, we only add its rect to the
collection if the region does not have a clip path. If a clip
path is present, we add it a collection of clip paths. The
final clip path is then set to a path consisting of all of the
discovered clip paths + the result of `pathWithShrinkWrappedRect`
for the collected rects.

* LayoutTests/interaction-region/clip-path-with-label-expected.txt: Added.
* LayoutTests/interaction-region/clip-path-with-label.html: Added.
* Source/WebCore/rendering/EventRegion.cpp:
(WebCore::EventRegionContext::shrinkWrapInteractionRegions):

Canonical link: https://commits.webkit.org/296646@main
